### PR TITLE
[Android] Paragraphs integration

### PR DESCRIPTION
**Pull request by jmartinesp**
## What?

Modify the Android code to handle paragraphs.

### Changes:
- Modified `SpanBackgroundHelper` so it can have a reference to the spanType and the text it's being applied to. This is useful to draw block spans with indentation for lists, i.e.. Also, clear position cache when the text inside the `TextView` is replaced.
- Fix issues with selecting indexes outside the TextView contents (`index >= textView.text.length()`).
- Fix rendering issues in `OrderedListSpan` when a list item has several lines.
- Re-work how HTML is parsed in `HtmlToSpansParser`, since the tags are parsed from children to parents, the result spans were added in reversed order and rendered wrong. Now they are added in order.
- Added line breaks between block elements when needed (paragraphs, list & items, code blocks, quotes...).
- Added NSBP/ZWSP extra characters to be able to render empty block elements, as spans with 0 length won't be rendered.
- Added indent/unindent actions.

## Why?

[`PSU-1078`](https://element-io.atlassian.net/browse/PSU-1078).

## Screenshots

image

### CODE REVIEW

**Review by jonnyandrew — COMMENTED** [review]
It works great! With only a few minor bugs.

The complexity of `HtmlToSpansParser` has gone up quite a bit after this change so I've added a few suggestions

File: platforms/android/library/src/main/java/io/element/android/wysiwyg/EditorEditText.kt [hunk 1]
```diff
@@ -236,6 +234,9 @@ class EditorEditText : TextInputEditText {
             return super.setText(text, type)
         }
 
+        inlineCodeBgHelper.clearCachedPositions()
+        codeBlockBgHelper.clearCachedPositions()
```
[the remark this answers is not in the document]
Comment (jmartinesp) [reply platforms/android/library/src/main/java/io/element/android/wysiwyg/EditorEditText.kt:238]: Oops, you're right. Nice catch!

File: platforms/android/library/src/main/java/io/element/android/wysiwyg/inlinebg/BlockRenderer.kt [hunk 2]
```diff
@@ -19,12 +22,19 @@ internal class BlockRenderer(
...
-        endOffset: Int
+        endOffset: Int,
+        text: Spanned,
+        spanType: Class<*>,
     ) {
+        val startIndex = layout.getOffsetForHorizontal(startLine, 0f)
+        val endIndex = layout.getOffsetForHorizontal(endLine, 0f)
+        val leadingMarginSpans = text.getSpans<LeadingMarginSpan>(startIndex, endIndex)
+            .filter { !spanType.isInstance(it) }
+        val leadingMargin = leadingMarginSpans.sumOf { it.getLeadingMargin(true) }
```
[the remark this answers is not in the document]
Comment (jmartinesp) [reply platforms/android/library/src/main/java/io/element/android/wysiwyg/inlinebg/BlockRenderer.kt:33]: I had my doubts about doing this, but for sure it'll be more performant. I guess with `clearCachedPositions` in place, there shouldn't be any issues.

[the remark this answers is not in the document]
Comment (jmartinesp) [reply platforms/android/library/src/main/java/io/element/android/wysiwyg/inlinebg/BlockRenderer.kt]: [`3603eb4`](https://github.com/matrix-org/matrix-rich-text-editor/pull/507/commits/3603eb4be192fd5956564c000cd9fc699862453f)

File: platforms/android/library/src/main/java/io/element/android/wysiwyg/EditorEditText.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmartinesp) [reply platforms/android/library/src/main/java/io/element/android/wysiwyg/EditorEditText.kt:238]: [`3603eb4`](https://github.com/matrix-org/matrix-rich-text-editor/pull/507/commits/3603eb4be192fd5956564c000cd9fc699862453f) (sorry, it got mixed with some other fixes)

File: platforms/android/library/src/main/java/io/element/android/wysiwyg/utils/HtmlToSpansParser.kt [hunk 3]
```diff
@@ -100,110 +143,128 @@ internal class HtmlToSpansParser(
...
                 }
+                addPlaceHolderSpan(newItem)
                 text.setSpan(newItem, text.length, text.length, Spanned.SPAN_INCLUSIVE_EXCLUSIVE)
             }
             "pre" -> {
-                text.setSpan(CodeBlock, text.length, text.length, Spanned.SPAN_INCLUSIVE_EXCLUSIVE)
+                addLeadingLineBreakIfNeeded(text.length)
+                val placeholder = PlaceholderSpan.CodeBlock()
+                addPlaceHolderSpan(placeholder)
+                text.setSpan(placeholder, text.length, text.length, Spanned.SPAN_INCLUSIVE_EXCLUSIVE)
```
[the remark this answers is not in the document]
Comment (jmartinesp) [reply platforms/android/library/src/main/java/io/element/android/wysiwyg/utils/HtmlToSpansParser.kt]: I tried to fix the 3 comments about the pending spans in [`637c9d9`](https://github.com/matrix-org/matrix-rich-text-editor/pull/507/commits/637c9d9abad31ee7e3fc7822829df0823353f9e5).

**Review by jonnyandrew — APPROVED** [review]
Nice one! And thanks for the changes.

**Pull request merged**